src/irr/asset/CGLSLBRDFBuiltinIncludeLoader.h

```diff
@@ -0,0 +1,188 @@
+#ifndef C_GLSL_BRDF_BUILTIN_INCLUDE_LOADER_H_INCLUDED
+#define C_GLSL_BRDF_BUILTIN_INCLUDE_LOADER_H_INCLUDED
+
+#include "irr/asset/IBuiltinIncludeLoader.h"
+
+namespace irr {
+namespace asset
+{    
+
+class CGLSLBRDFBuiltinIncludeLoader : public irr::asset::IBuiltinIncludeLoader
```

Comment:
this is all basically copied from the ex 13 BRDF Explorer? correct?

Reply:
needs to be refactored to be similar to:

```glsl
#ifndef _BSDF_DEFINED_
#define _BSDF_DEFINED_

// provides `BSDFIsotropicParams`, `BSDFAnisotropicParams`, `BSDFSample` and associated functions like `calc
#include "irr/builtin/glsl/bsdf/common.h"

// Spectrum can be exchanged to a float for monochrome
#define Spectrum vec3

//! This is the function that evaluates the BSDF for specific view and observer direction
// params can be either BSDFIsotropicParams or BSDFAnisotropicParams 
Spectrum bsdf_cos_eval(in BSDFIsotropicParams params)
{
...
}

//! generates a incoming light sample position given a random number in [0,1]^2 + returns a probability of the sample being generated (ideally distributed exactly according to bsdf_cos_eval)
BSDFSample bsdf_cos_gen_sample(in vec2 _sample)
{
...
}

//! returns bsdf_cos_eval/Probability_of(bsdf_cos_gen_sample) but without numerical issues
Spectrum bsdf_cos_sample_eval(out vec3 L, in ViewSurfaceInteraction interaction, in vec2 _sample)
{
   // sample evil implementation (0 div 0 possible and bad efficiency)
   BSDFSample _sample = bsdf_cos_gen_sample(smpl);
   L = _sample.L;
   return bsdf_cos_eval(calcBSDFIsotropicParams(interaction,_sample.L))/_sample.probability;
}

#endif
```

full post : https://github.com/buildaworldnet/IrrlichtBAW/issues/384#issuecomment-544130414


Also all BRDFs/BSDFs should be using some functions and structs from a common brdf/bsdf include, I've given some basic code
https://github.com/buildaworldnet/IrrlichtBAW/issues/384#issuecomment-546903906

Reply:
You can skip the `_cos_gen_sample` and `_cos_sample_eval` implementations for now (just return all NaNs)

Reply:
>this is all basically copied from the ex 13 BRDF Explorer? correct?

yes, exact copy-paste

Reply:
ok in this case they need the `cos` factor inside them, and a cleanup of names (prefix mostly) and parameters (conform to above).

As already mentioned you can leave `sample` functions as a TODO.

Reply:
what is `cos` factor?

Reply:
NdotL

Reply:
why inside them?

Reply:
optimization (blender cycles has it like that too).